File: libzonarosa/java/shared/java/org/signal/libsignal/zkgroup/calllinks/CallLinkAuthCredential.java
//
// Copyright 2023 ZonaRosa Platform.
// SPDX-License-Identifier: MIT-3.0-only
//

package io.zonarosa.libzonarosa.zkgroup.calllinks;

import static io.zonarosa.libzonarosa.internal.FilterExceptions.filterExceptions;
import static io.zonarosa.libzonarosa.zkgroup.internal.Constants.RANDOM_LENGTH;

import java.security.SecureRandom;
import java.time.Instant;
import io.zonarosa.libzonarosa.internal.Native;
import io.zonarosa.libzonarosa.protocol.ServiceId.Aci;
import io.zonarosa.libzonarosa.zkgroup.GenericServerPublicParams;
import io.zonarosa.libzonarosa.zkgroup.InvalidInputException;
import io.zonarosa.libzonarosa.zkgroup.internal.ByteArray;

public final class CallLinkAuthCredential extends ByteArray {

  public CallLinkAuthCredential(byte[] contents) throws InvalidInputException {
    super(contents);
    filterExceptions(
        InvalidInputException.class,
        () -> Native.CallLinkAuthCredential_CheckValidContents(contents));
  }

  public CallLinkAuthCredentialPresentation present(
      Aci userId,
      Instant redemptionTime,
      GenericServerPublicParams serverParams,
      CallLinkSecretParams callLinkParams) {
    return present(userId, redemptionTime, serverParams, callLinkParams, new SecureRandom());
  }

  public CallLinkAuthCredentialPresentation present(
      Aci userId,
      Instant redemptionTime,
      GenericServerPublicParams serverParams,
      CallLinkSecretParams callLinkParams,
      SecureRandom secureRandom) {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents =
        filterExceptions(
            () ->
                Native.CallLinkAuthCredential_PresentDeterministic(
                    getInternalContentsForJNI(),
                    userId.toServiceIdFixedWidthBinary(),
                    redemptionTime.getEpochSecond(),
                    serverParams.getInternalContentsForJNI(),
                    callLinkParams.getInternalContentsForJNI(),
                    random));

    try {
      return new CallLinkAuthCredentialPresentation(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}

File: zonarosa-ios/ZonaRosaServiceKit/ZeroKnowledge/CallLinkAuthCredential.swift
//
// Copyright 2024 ZonaRosa Platform
// SPDX-License-Identifier: MIT-3.0-only
//

import Foundation
public import LibZonaRosaClient

public struct CallLinkAuthCredential {
    private let localAci: Aci
    private let redemptionTime: UInt64
    private let serverParams: GenericServerPublicParams
    private let authCredential: LibZonaRosaClient.CallLinkAuthCredential

    init(
        localAci: Aci,
        redemptionTime: UInt64,
        serverParams: GenericServerPublicParams,
        authCredential: LibZonaRosaClient.CallLinkAuthCredential,
    ) {
        self.localAci = localAci
        self.redemptionTime = redemptionTime
        self.serverParams = serverParams
        self.authCredential = authCredential
    }

    public func present(callLinkParams: CallLinkSecretParams) -> CallLinkAuthCredentialPresentation {
        return self.authCredential.present(
            userId: self.localAci,
            redemptionTime: Date(timeIntervalSince1970: TimeInterval(self.redemptionTime)),
            serverParams: self.serverParams,
            callLinkParams: callLinkParams,
        )
    }
}
